File: src/forward_model/radiative_transfer/dispatcher.zig
const std = @import("std");
const common = @import("root.zig");
const labos = @import("labos/root.zig");

pub fn prepare(request: common.DispatchRequest) common.PrepareError!common.Route {
    return common.prepareRoute(request);
}

pub fn executePrepared(
    allocator: std.mem.Allocator,
    route: common.Route,
    input: common.ForwardInput,
) common.ExecuteError!common.ForwardResult {
    return executePreparedWithLabosWorkspace(allocator, route, input, null);
}

pub fn executePreparedWithLabosWorkspace(
    allocator: std.mem.Allocator,
    route: common.Route,
    input: common.ForwardInput,
    workspace: ?*labos.Workspace,
) common.ExecuteError!common.ForwardResult {
    if (route.regime != .nadir) return common.Error.UnsupportedObservationRegime;
    if (route.execution_mode != .scalar) return common.Error.UnsupportedExecutionMode;
    if (route.derivative_mode == .numerical) return common.Error.UnsupportedDerivativeMode;
    return labos.executeWithWorkspace(allocator, route, input, workspace);
}

pub fn execute(
    allocator: std.mem.Allocator,
    request: common.DispatchRequest,
    input: common.ForwardInput,
) common.ExecuteError!common.ForwardResult {
    const route = try prepare(request);
    return executePrepared(allocator, route, input);
}

File: rust_src/forward_model/radiative_transfer/dispatcher.rs
use crate::{
    forward_model::radiative_transfer::{
        self, DispatchRequest, Error, ForwardInput, ForwardResult, Result, Route,
    },
    input::{DerivativeMode, ObservationRegime},
};

pub fn prepare(request: DispatchRequest) -> Result<Route> {
    radiative_transfer::prepare_route(request)
}

pub fn execute_prepared(route: Route, input: &ForwardInput) -> Result<ForwardResult> {
    if route.regime != ObservationRegime::Nadir {
        return Err(Error::UnsupportedObservationRegime);
    }
    if route.execution_mode != radiative_transfer::ExecutionMode::Scalar {
        return Err(Error::UnsupportedExecutionMode);
    }
    if route.derivative_mode == DerivativeMode::Numerical {
        return Err(Error::UnsupportedDerivativeMode);
    }
    radiative_transfer::labos::execute(route, input)
}

pub fn execute(request: DispatchRequest, input: &ForwardInput) -> Result<ForwardResult> {
    let route = prepare(request)?;
    execute_prepared(route, input)
}
